Feature/persistent configuration

Hamer2 opened the pull request:
Completed up to task 3 from the issues section. I haven't checked too hard for bugs but it seems to be working so far.
You can save, reset and back out of the config menu and it will save your changes to a config.json file in the root directory. If no json file exists it will create one as well.

Had to make changes to /controller/configuration to refactor the code. Also created an Fxml loader factory and changed /helpers/ScenesManager to add that in.

silentlie committed:
Merge pull request #35 from silentlie/dev

fix: game logic clear lines

Hamer2 committed:
feat: added configuration record

Hamer2 committed:
feat: added configuration record

Hamer2 committed:
feat: implemented json configuration service with atomic writes and default fallback

Hamer2 committed:
misc: added comments to JsonConfigurationService

Hamer2 committed:
feat: working configuration persistence

Details:

Hamer2 committed:
Feature:updated configuration.fxml for dynamic player fields

Hamer2 committed:
misc: added comments to ConfigurationController

Hamer2 commented:
Hey @silentlie, I have completed the persistent config feature, the settings now load from config.json, can be updated via the configuration page and are saved back with validation and atomic writes. At the moment the controller handles persistence using a cached copy of the config, not through AppContext. Once the AppContext is done/merged we should be able to hook into that. Let me know if there are any issues as well.

silentlie commented on src/main/java/com/example/tetris/models/Configuration.java:
```diff
@@ -0,0 +1,73 @@
+    if (player1 == null || player2 == null) {
+      throw new IllegalArgumentException("Player type must not be null");
+    }
+  }
```
Instead of throwing errors, maybe use default values instead

silentlie commented on src/main/resources/com/example/tetris/configuration.fxml:
Please use Gridpane for responsive compatibility:
3 Columns
10 Rows
Using percentWidth and percentHeight so the view can be responsive
Using GridPane.columnSpan to span across multiple columns
Like the Header should span across 3 columns
Or Sliders should span across 2 columns

silentlie commented on config.json:
Add config.json to .gitignore

silentlie commented on src/main/java/com/example/tetris/helpers/ScenesManager.java:
Please merge `dev` into this branch to continue

silentlie commented on src/main/java/com/example/tetris/helpers/FxmlLoaderFactory.java:
After merging `dev` into this branch, access the config cache and service from the app context instead, though setConfigurationService

silentlie commented on src/main/java/com/example/tetris/controllers/ConfigurationController.java:
```diff
@@ -0,0 +1,177 @@
+  @FXML private javafx.scene.control.ChoiceBox<PlayerType> player2Choice;
+
+  // Container for player 2 controls so they can be enabled/disabled as a group
+  @FXML private javafx.scene.layout.HBox player2Box;
```
Use import instead of using dirrectly

silentlie replied:
You may not need player2Box

silentlie commented on src/main/java/com/example/tetris/controllers/ConfigurationController.java:
```diff
@@ -0,0 +1,177 @@
+    // Set slider ranges
+    widthSlider.setMin(5); widthSlider.setMax(15);
+    heightSlider.setMin(15); heightSlider.setMax(30);
+    levelSlider.setMin(1); levelSlider.setMax(10);
```
Set in FXML already

silentlie commented on src/main/java/com/example/tetris/controllers/ConfigurationController.java:
```diff
@@ -0,0 +1,177 @@
+    cachedConfig = service.load();
+
+    // Populate player choice boxes with enum values
+    player1Choice.getItems().setAll(PlayerType.values());
+    player2Choice.getItems().setAll(PlayerType.values());
```
Maybe enum should have a label field instead of all capital from enum

silentlie commented on src/main/java/com/example/tetris/controllers/ConfigurationController.java:
```diff
@@ -0,0 +1,177 @@
+            heightSlider.valueProperty()));
+    levelValue.textProperty().bind(
+        Bindings.createStringBinding(() -> String.valueOf((int) levelSlider.getValue()),
+            levelSlider.valueProperty()));
```
Should use IntBinding

silentlie commented on src/main/java/com/example/tetris/controllers/ConfigurationController.java:
```diff
@@ -0,0 +1,177 @@
+  private void handleReset() {
+    Configuration defaults = new Configuration();
+    setFormFromConfig(defaults);
+    service.save(defaults);
```
You don't actually need to save here

silentlie commented on src/main/java/com/example/tetris/services/JsonConfigurationService.java:
Later, there will be JsonUtil, so the services can use

silentlie reviewed (changes requested):
Overall looks good, but need to merge `dev` into this branch to integrate new update from `dev`, also please check the comments.

silentlie committed:
Merge branch 'dev' into feature/persistent-configuration

# Conflicts:
#	src/main/java/com/example/tetris/controllers/ConfigurationController.java
#	src/main/java/com/example/tetris/helpers/ScenesManager.java
#	src/main/resources/com/example/tetris/configuration.fxml

silentlie committed:
merge `dev` in this branch, resolve conflict and format
but have not resolved ConfigurationController start with loader, please use AppContext instead.

silentlie commented:
@Hamer2, I have merged `dev` into this branch. Please make a git pull to get the update and continue working from here.
If you don't understand anything, please tag me

silentlie committed:
Merge branch 'dev' into feature/persistent-configuration

# Conflicts:
#	src/main/java/module-info.java

silentlie committed:
fix: docs in JsonUtil

Hamer2 committed:
feat: Merge, AppContext persistence, UI fix, clean up

Hamer2 committed:
feat: Merge, AppContext persistence, UI fix, clean up

Hamer2 committed:
feat: Merge, AppContext persistence, UI fix, clean up, integrate JsonUtil

Hamer2 commented:
@silentlie merged and added your feedback.

silentlie commented on src/main/java/com/example/tetris/controllers/ConfigurationController.java:
```diff
@@ -0,0 +1,150 @@
+    // toggle P2 availability with extend mode
+    extendedCheck.selectedProperty().addListener((obs, oldVal, newVal) -> {
+      player2Choice.setDisable(!newVal);
+      if (!newVal) player2Choice.setValue(PlayerType.OFF);
```
Adding PlayerType.OFF, later can cause bugs if both players are OFF, intial design is fine

silentlie commented on src/main/java/com/example/tetris/controllers/ConfigurationController.java:
```diff
@@ -0,0 +1,150 @@
+  private AppContext ctx;
+  private ConfigurationService service;
+  private ConfigurationCache cache;
+  private Configuration cachedConfig;
```
Should not create a copy/refrence here, access dirrectly from AppContext

silentlie commented on src/main/java/com/example/tetris/controllers/ConfigurationController.java:
```diff
@@ -0,0 +1,150 @@
+
+  private void handleBack() {
+    Configuration current = formToConfig();
+    if (!current.equals(cachedConfig)) {
```
compare with the ctx cache instead

silentlie commented on pom.xml:
Duplicate jackson

silentlie reviewed (changes requested):
Tiny fixes and you're good to go.

Hamer2 committed:
fix: updated ConfigurationController to only use AppContext, removed duplicate jackson from pom, removed PlayerType.OFF

silentlie reviewed (dismissed):
Good works

silentlie reviewed (approved):
Good works
